nutils/points.py

```diff
@@ -0,0 +1,442 @@
+# Copyright (c) 2014 Evalf
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+'''
+The points module defines the :class:`Points` base class, which bundles point
+coordinates, point weights, a local triangulation and a hull triangulation. The
+module provides several different implementations such as :class:`TensorPoints`
+and :class:`SimplexGaussPoints` that reflect the variety of elements in the
+:mod:`nutils.element` module.
+'''
+
+from . import types, transform, numeric, util
+import numpy, functools, itertools, warnings
+_ = numpy.newaxis
+
+class Points(types.Singleton):
+  '''Collection of points on an element.
+
+  The :class:`Points` base class bundles point coordinates, point weights,
+  a local triangulation and hull triangulation. Of these only the coordinates
+  are mandatory, and should be provided by the derived class in the form of the
+  ``coords`` attribute. Of the remaining properties only :func:`hull` has a
+  functional base implementation that relies on the availability of ``tri``.
+  '''
+
```

Comment:
You could add

    .. attribute:: coords

      Coordinates of the points as a :class:`float` array.

    .. attribute:: weights

      Weights of the points as a :class:`float` array.

to the docstring and refer to the attributes (`coords`, `hull`, `tri`) with `` :attr:`...` ``.